app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,34 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+  def facebook
+    auth = request.env['omniauth.auth']
+
+    if auth.info.email.blank?
+      redirect_to user_omniauth_authorize_path(
+        :facebook,
+        auth_type: 'rerequest',
+        scope: 'email,public_profile'
+      )
+    else
+      user = User.find_by(provider: auth.provider, uid: auth.uid)
+
+      if user.nil?
+        user = User.find_by(email: auth.info.email)
+
+        if user.nil?
+          user = User.create_from_omniauth(auth)
+        elsif user.uid != auth.uid
+          user.update_from_omniauth(auth)
+        end
+      end
```

Comment:
@stedmanblake I cleaned up this controller a bit to keep it a bit simpler. The main things I'm checking for here are:

1. An OAuth user with the provider / uid – since they likely hit the "Sign In with FB" link, would like to look up that specific case first. E.g. we only issue on query to sign someone in who's already got an OAuth connected account.

2. If there is no OAuth user, then issue a second query to find an email user – if this email user exists, update their record with the appropriate provider/uid.

3. If there is both no OAuth user nor an email user, issue the third query and create the account.